### PR TITLE
Change API to support specifying "single", "optional" and "variadic" input

### PR DESCRIPTION
**Pull request by linkerzhang**
This is a follow-up of https://github.com/onnx/onnx/issues/285.

More thoughts and open questions:
1. Remove all "NumInputs", "NumOutputs" API, which seem to be not needed now given we have "optional", "variadic" types specified clearly.
2. Do we need to support "optional" and "variadic" outputs? for example, operator "Split". How do we hook up the outputs of "Split" with another node (op)?

### CODE REVIEW

File: onnx/defs/schema.cc [hunk 1]
```diff
@@ -49,8 +49,8 @@ const std::string& OpSchema::FormalParameter::GetDescription() const {
   return description_;
 }
 
-bool OpSchema::FormalParameter::IsOptional() const {
-  return is_optional_;
+OpSchema::FormalParameterOption OpSchema::FormalParameter::GetOption() const {
+  return param_option_;
 }
 
 void OpSchema::Verify(const NodeProto& node) const {
```
Comment (gramalingam) [inline onnx/defs/schema.cc]: I think that the check if node.input_size() > max_input should be made only if there is no variadic parameter. (Of course, this depends on what max_input is set to for an operator with a variadic parameter.)

File: onnx/defs/schema.cc [hunk 2]
```diff
@@ -110,11 +110,11 @@ void OpSchema::Verify(const NodeProto& node) const {
 
   // Check the values of inputs / outputs
   for (int in_idx = 0; in_idx < node.input_size(); ++in_idx) {
-    if (node.input(in_idx).empty() && !(inputs_[in_idx].IsOptional())) {
+    if (node.input(in_idx).empty() && (Single == inputs_[in_idx].GetOption())) {
```
Comment (gramalingam) [inline onnx/defs/schema.cc]: Do we want to permit only the last parameter to be variadic? If we allow a parameter other than the last one to be variadic (still restricting the number of variadic parameters to be one), then the association between the index of an actual parameter and the index of the corresponding formal parameter is more complex. For parameters before the optional one, it is a direct match. For parameters after the optional one, we need to count from the back ... that is, the k-th parameter from the end in the actual parameter list will correspond to the k-th parameter from the end in the formal parameter list.

**Comment by gramalingam** [comment]
Looks good to me (modulo the question about max_input and check on node.input_size above). I think it will be helpful to document the assumptions and restrictions: as I understand it,
(a) Only last parameter is allowed to be variadic.
(b) What is max_input supposed to be if we have k parameters and last is variadic? As you point out, it may be a good idea to eliminate NumInput(...) if it can be calculated. Then we don't have to worry about this.

**Comment by gramalingam** [comment]
The question about outputs (2) is a good one. It makes sense to do the same. We also need to define what the behavior is for several of these cases for the output:
(a) Can the user specify an empty-string for an output that they do not care about?
(b) If the user specifies a name, say X, for an output that doesn't have a value (e.g., DropOut may not return one of the output values depending on test-mode), then what is X bound to?

File: onnx/defs/schema.cc (hunk 1 above)
Comment (linkerzhang) [reply onnx/defs/schema.cc]: Exactly on "Of course, this depends on what max_input is set to for an operator with a variadic parameter."

That's why I put a more thought "Remove all "NumInputs", "NumOutputs" API, which seem to be not needed now given we have "optional", "variadic" types specified clearly." with this PR.

I'm thinking of remove the min/max number of input setting. Because, for single/optional input types, it's easy to get to know min/max without an explict set. For variadic, it's always 0-n setting.

Make sense?

If we all agree on removing the "NumInputs, NumOutputs" APIs, I'll do that and change the logic here accordingly too.

Comment (gramalingam) [reply onnx/defs/schema.cc]: Yes. I guess others can chime in if they see a need for NumInputs for any other future extension.

File: onnx/defs/math/defs.cc [hunk 3]
```diff
@@ -396,7 +396,7 @@ OPERATOR_SCHEMA(Max)
 Element-wise max of each of the input tensors. All inputs and outputs must
 have the same shape and data type.
 )DOC")
-    .Input(0, "data_0", "First of the input tensors.", "T")
+    .Input(0, "data_0", "First of the input tensors.", "T", OpSchema::Variadic)
```
Comment (dzhulgakov) [inline onnx/defs/math/defs.cc]: we should change "First" here and in other places below

File: onnx/defs/rnn/defs.cc [hunk 4]
```diff
@@ -37,11 +37,11 @@ std::function<void(OpSchema&)> RNNDocGenerator(const char* name) {
                      "Optional tensor specifying lengths of the sequences in a batch. "
                      "If not specified - assumed all sequences in the batch to have "
                      "length `seq_length`. It has shape `[batch_size]`.", "T1",
-                     true /*optional*/);
+                     OpSchema::Optional /*optional*/);
```
Comment (dzhulgakov) [inline onnx/defs/rnn/defs.cc]: nit: remove "optional" comment

File: onnx/defs/schema.cc (hunk 1 above)
Comment (dzhulgakov) [reply onnx/defs/schema.cc]: It might make sense to keep NumInputs just in case we'd need some very fancy operator versions (e.g. for training batchnorm there might be 1 or 3 arguments, but not 2; though it's a bit far-fetched).

But I agree with you that we can do Input() and Output() specs a default. E.g. if none of NumInputs() functions is invoked we just infer it from the Input() spec. Then we can first verify that all existing op specifications are valid before removing NumInput() specs.

But that can happen in a separate PR if your prefer.

File: onnx/defs/schema.cc (hunk 2 above)
Comment (dzhulgakov) [reply onnx/defs/schema.cc]: I'd say to allow variadic be at the end only, it makes it easier to reason (seems like you're already checking it below)

I also realized that this code had bug before, we need to check that `in_idx < inputs_.size()`

File: onnx/defs/schema.cc [hunk 5]
```diff
@@ -499,13 +499,19 @@ void OpSchema::Finalize() {
...
+  for (int i = 0; i < (int)(inputs_.size()) - 1; ++i) {
+    ENFORCE(Variadic != inputs_[i].GetOption());        
+  }
+
   // TODO: also cover checks for arbitrary number of inputs
   // allow extra tailing inputs not be present if all inputs at the end are
   // marked as optional
   if (max_input_ < std::numeric_limits<int>::max()) {
     int ind = max_input_;
     for (auto& input : inputs_) {
```
Comment (dzhulgakov) [inline onnx/defs/schema.cc]: this code should probably also look at indices from the end only. Original (my) implementation was not very correct

**Comment by buehrer** [comment]
I agree with the other comments, looks good.  Allowing only the last input to be variadic makes sense to me.

File: onnx/defs/schema.cc (hunk 1 above)
Comment (linkerzhang) [reply onnx/defs/schema.cc]: sure, will make that change in a separate PR.

File: onnx/defs/schema.cc (hunk 5 above)
Comment (linkerzhang) [reply onnx/defs/schema.cc]: will revisit this logic in separate PR together with generating min/max input/output without NumInputs, NumOutputs specification.
Basically, min is the number of "Single"s. max is sum of number of singles and optionals. if there's variadic, max is INT_MAX.
Make sense?

File: onnx/defs/math/defs.cc (hunk 3 above)
Comment (linkerzhang) [reply onnx/defs/math/defs.cc]: Yep. FIxed and Thank you!

File: onnx/defs/rnn/defs.cc (hunk 4 above)
Comment (linkerzhang) [reply onnx/defs/rnn/defs.cc]: fixed. Thank you!
